src/solver/opamSolver.ml

```diff
@@ -772,19 +772,27 @@ let dump_universe universe oc =
 
 let filter_solution filter t =
   let t = OpamCudf.ActionGraph.copy t in
-  let rec rm iter_deps v =
-    if OpamCudf.ActionGraph.mem_vertex t v then (
-      iter_deps (rm iter_deps) t v;
+  let rec rm ?(check=fun _ -> true) iter_deps v =
+    if OpamCudf.ActionGraph.mem_vertex t v && check v then (
+      iter_deps (rm ~check iter_deps) t v;
       OpamCudf.ActionGraph.remove_vertex t v
     ) in
   OpamCudf.ActionGraph.iter_vertex
-    (function
-      | `Remove nv as a when not (filter (OpamCudf.cudf2opam nv)) ->
-        rm OpamCudf.ActionGraph.iter_pred a
-      | (`Install nv | `Change (_,_,nv)) as a
-        when not (filter (OpamCudf.cudf2opam nv)) ->
-        rm OpamCudf.ActionGraph.iter_succ a
-      | _ -> ())
+    (fun a ->
+       match a with
```

Comment:
```suggestion
    (function
```
`a` is not used anywhere (it’s all redefined by the `as a` below) and this would avoid the change of indentation